### ISSUE TITLE
Rocket Analytics blocks page generation: wp_remote_post() to mixpanel-proxy.group.one with no blocking => false adds 4-16s TTFB when that endpoint is degraded

### ISSUE THREAD

**Issue by gconkie** [wp-media/wp-rocket#8669]
# Rocket Analytics blocks page generation: `wp_remote_post()` to `mixpanel-proxy.group.one` with no `blocking => false` adds 4–16s TTFB when that endpoint is degraded

**WP Rocket version:** 3.23.1 (also present in 3.22.0.1)
**WordPress:** 7.0.2 · **PHP:** 8.0 – 8.3 · **Server:** LiteSpeed / CloudLinux (shared hosting)
**Observed:** 3 August 2026

## Summary

`WPMedia\Mixpanel\WPConsumer::persist()` sends analytics batches with a **blocking** `wp_remote_post()`. WordPress defaults `blocking` to `true`, and no `'blocking' => false` is passed. When `mixpanel-proxy.group.one` is slow or failing, **every uncached page load stalls for as long as that endpoint takes to fail.**

We hit this on a shared-hosting estate: the endpoint was returning 502/504 after 4–16s, which added ~16s to TTFB on every cache miss across 39 WordPress installs. Disabling Rocket Analytics took affected sites from ~16.5s to ~0.6s.

## The code

`vendor/wp-media/wp-mixpanel/src/WPConsumer.php`:

```php
$response = wp_remote_post(
    $url,
    [
        'timeout' => $this->timeout,
        'body'    => $data,
    ]
);
```

`vendor/wp-media/wp-mixpanel/src/Tracking.php`:

```php
const HOST = 'mixpanel-proxy.group.one';
```

There is no `'blocking' => false`, so the request is synchronous and the page waits on it.

## Evidence

Captured with a temporary mu-plugin hooking `pre_http_request` / `http_api_debug`, on an uncached front-end request:

```
16.01s  blocking=yes  timeout=30  by=wp-rocket  504  ->  https://mixpanel-proxy.group.one/engage
 2.42s  blocking=yes  timeout=30  by=wp-rocket  502  ->  https://mixpanel-proxy.group.one/engage
```

Independently, from the same server, the endpoint itself:

```
502 in 5.08s
502 in 4.21s
504 in 15.11s
```

`mixpanel-proxy.group.one` resolves to `5.249.224.2` (PTR `lb.pub.k8spod3-cph3.g1i.one`) and presents a default **"Kubernetes Ingress Controller Fake Certificate"** for `ingress.local` on a direct connection.

## Impact

Same page, cache-busted so it always renders, before and after setting `wp_rocket_settings['analytics_enabled'] = 0`:

```
before:  16.47s  16.52s  16.56s  16.84s
after:    0.54s   0.57s   0.58s   0.61s
```

Roughly **27×**. Across the estate, affected sites measured 16–64s uncached before the change and 1–4s after.

**This is easy to miss**, which is why it ran unnoticed for a long time: page caching hides it from ordinary visitors. Our worst-affected site returned `200 OK in 48ms` from cache while *every* uncached request took 16s. It only surfaces on a cache miss, first hit after a purge, a logged-in user, an uptime monitor that cache-busts, or a crawler walking uncached URLs. Under crawler load it is materially worse, because each miss holds a PHP worker for the full timeout.

## Reproduction

1. WP Rocket with Rocket Analytics enabled.
2. Make `mixpanel-proxy.group.one` slow or unreachable (blackhole `5.249.224.2`, or observe during a real degradation).
3. Request any URL that is not in the page cache.
4. TTFB increases by the endpoint's time-to-failure.

## Why I think this is a bug rather than tuning

- **Analytics is non-essential.** A telemetry POST should never be able to delay page generation for a visitor.
- **`timeout` defaults high.** We observed `timeout=30`, six times the 5s you settled on for licence checks.
- **There is a direct precedent in this repo.** #6715 ("Avoid requests to wp-rocket.me to negatively impact site performances during downtime"), raised after the June 2024 outage, describes exactly this failure mode for licence/update checks:

  > "When WP Rocket (website) is down or slow, which can unfortunately happen, this creates slowness on every website checking for updates and license validity. The requests go in timeout (5 seconds)."

  That was fixed in #6739 with caching, a retry backoff and a 5s timeout, but the fix was scoped to licensing. The same reasoning applies to analytics, and analytics is the weaker justification of the two for ever blocking.

## Suggested fix

1. **`'blocking' => false`** on the analytics POST. Fire-and-forget is sufficient for telemetry.
2. **Short timeout** (1–3s) with a filter, rather than the current default.
3. **Circuit breaker** - after N consecutive failures, stop attempting for a cool-off period, as done for licensing in #6739.
4. **Keep it off the front-end render path entirely** - queue events and flush on cron / Action Scheduler / admin requests, never during a visitor's uncached page build.

(1) alone removes the visitor-facing impact; (3) and (4) prevent it recurring in another form.

## Workaround for anyone hitting this

Disable Rocket Analytics (Settings → Analytics, or `wp_rocket_settings['analytics_enabled'] = 0`). Note that a queue of pending events may still flush on the first few requests after disabling, so the improvement is not instantaneous.

---

Reported by [SmartHost.ie](https://smarthost.ie). Happy to supply further captures, or to test a patched build against the failing endpoint. We still have the reproduction environment.

**Comment by johan-las** [comment]
related: 

- https://onecomhelp.zendesk.com/agent/tickets/18224352
- https://onecomhelp.zendesk.com/agent/tickets/18236615

**Comment by juricazuanovic** [comment]
Related: https://onecomhelp.zendesk.com/agent/tickets/18228986

**Comment by gsusI** [comment]
Independent production confirmation from a WooCommerce store running WP Rocket 3.23.1, WordPress 7.0.2, PHP 8.4.23, and Nginx/PHP-FPM on Cloudways.

This caused a full availability incident, not merely slower cache misses. New Relic captured:

- A Store API transaction lasting 77.18 seconds.
- Three `mixpanel-proxy.group.one/engage` calls consuming 75.3 seconds—97.56% of the transaction.
- Blocking spans of 30.00s and 15.29s.
- The call chain `MixpanelBaseProducer::flush()` → `WPConsumer::persist()` → `wp_remote_post()` → `Curl::request()`.
- Mixpanel consumed 88.73% of `index.php` wall time. PHP averaged 1.89s; Redis and MySQL took milliseconds.

These calls saturated a 20-child PHP-FPM pool and caused Nginx upstream connection errors.

We disabled telemetry with:

```bash
wp option update rocket_mixpanel_optin 0
wp cache flush
```

Post-fix:

- Zero further Group One spans.
- `index.php`: 1.84s average, 4.341s p95/max.
- `admin-ajax.php`: 2.15s average, 4.078s p95.
- Direct-origin Store API requests: 0.95–1.42s sequential.

Strong +1 for keeping analytics entirely off the visitor request path. A performance plugin’s telemetry must fail open. Queueing through cron with a circuit breaker would be safer than synchronous requests.

I can provide anonymized New Relic screenshots or additional traces if useful.

**Comment by johan-las** [comment]
related: https://onecomhelp.zendesk.com/agent/tickets/18242654

**Comment by nightjoker7** [comment]
This is not a "medium" issue and it should not still be open without a fix. I want to put hard numbers on it, because I think the severity has been badly misjudged.

## What the defect actually is

`WPConsumer::persist()` (`vendor/wp-media/wp-mixpanel/src/WPConsumer.php`) calls:

```php
$response = wp_remote_post(
    $url,
    [
        'timeout' => $this->timeout,   // 30
        'body'    => $data,
    ]
);
```

No `'blocking' => false`. A **30-second** timeout. Host is hardcoded at `Tracking.php:9` — `const HOST = 'mixpanel-proxy.group.one'`.

That is a fire-and-forget usage-analytics ping, executed **synchronously, inside the front-end page render path, in a performance plugin**. Whatever the endpoint's uptime is, that construction is wrong on first principles. The outage did not create this defect. It only revealed it.

## Measured impact on a production WooCommerce store

Measured directly from my origin server on 2026-08-04, Cloudflare bypassed:

```
mixpanel-proxy.group.one
  run1  404  TTFB= 0.56s
  run2  504  TTFB=10.57s
  run3  404  TTFB=15.48s
  run4  404  TTFB=15.51s
  run5  404  TTFB=15.50s
```

Socket-level capture of a single uncached page render (111,687 bytes total):

```
[ 0.575s]  98,986 bytes delivered   <- 89% of the page. Page is VISIBLE to the user.
[30.238s]  final 12,701 bytes + connection close
```

Cold, cache-busted page loads before vs. after neutralising this one call:

| Page | Before | After |
|---|---|---|
| `/` | 16.3s / 36.6s / 46.2s | 0.74–0.93s |
| `/bcm-guide/` | no response at 25s | 0.68s |
| `/ecm-guide/` | 17.8s | 0.69s |
| `/blog/` | 16.6s | 0.87s |
| `/shop/` | 6.8s | 1.11s |

Some requests returned nothing at all. On a storefront.

## The part that makes this severe rather than merely slow

**Two WP Rocket features compounded into total input death, and neither of them logged anything.**

Because the document never finishes, `window.load` never fires. And WP Rocket's own **Delay JavaScript Execution** installs a capture-phase handler on `window` that calls `preventDefault()`, `stopPropagation()` and `stopImmediatePropagation()` on **every click**, queues it, and replays the queue only at the very end of its chain:

```js
await this.C(this.delayedScripts.normal);
await this.C(this.delayedScripts.defer);
await this.C(this.delayedScripts.async);
await this.T();
await this.F();
await this.j();   // <-- waits on the real window 'load' event
await this.A();
…
this.U();         // <-- only here are the user's queued clicks replayed
```

`j()` is `new Promise(t => window.addEventListener("load", t))`. **There is no timeout anywhere in that chain, and no fallback.**

So the real-world behaviour for a visitor was:

1. Page renders and looks completely normal.
2. They tap a navigation link. Nothing happens.
3. They tap again. Nothing happens. Repeatedly.
4. No error. No console message. No spinner state that suggests anything is wrong.
5. Closing the browser window and reopening it "fixes" it — because the retry hits a warm cache and skips the render entirely.

I verified this in a browser on a fully loaded, idle page: a click was still returning `defaultPrevented: true`.

That is not a slow site. That is a site that is up, looks fine, takes money on some pages, and **silently ignores customers trying to navigate or check out.** Anyone who was not sitting on a packet capture would conclude their own theme was broken. I nearly did.

## The response is the actual problem

- **Aug 3** — issue opened, with a correct diagnosis and four concrete fixes handed to you by a customer.
- **Aug 3** — assigned.
- Since then: no maintainer comment, no label, no milestone, no linked PR, no release, no advisory.

Meanwhile:

- A **third-party host (BigScoots) shipped a mitigation to its whole fleet within 16 minutes** of identifying it (`status.bigscoots.com/incident/49`), by null-routing your telemetry domain in `/etc/hosts`. Their status page named WP Rocket as the cause.
- Your paying customers diagnosed your bug, wrote your patch options, and mitigated it themselves.
- There has been **no email, no dashboard notice, and no status page entry** telling affected licence holders that their sites were degraded by this plugin. Most of them still do not know. Their stores were quietly losing conversions and they were blaming their hosts, their themes, or their developers.

The fix for the acute phase is one array key. `'blocking' => false`. That is a same-day hotfix in a plugin that ships releases regularly. Choosing not to ship it — while an external hosting company shipped a workaround for you — is the part that is hard to accept from a paid product.

And this is a **paid** product. The entire value proposition is page speed. A defect in this plugin that adds 16–46 seconds to uncached responses, and disables navigation outright, inverts the only thing customers buy it for.

## What I am asking for

1. **Ship `'blocking' => false` as a hotfix release now.** Do not wait for the endpoint to recover, and do not bundle it into a feature release.
2. **Add a hard timeout (1–3s) plus a circuit breaker** so consecutive failures stop further attempts for a cooldown period.
3. **Move analytics off the front-end render path entirely.** It belongs in a cron/Action Scheduler job. There is no justification for a usage-metrics call ever running while a visitor waits for HTML.
4. **Re-label this issue to critical** and state a target release publicly in this thread.
5. **Notify licence holders.** An email and a dashboard notice naming the symptom — "navigation links may not respond; uncached pages may take 15–45s" — so people stop hunting ghosts in their own code.
6. **Audit every other outbound HTTP call in the plugin** for the same pattern and post the results. If telemetry shipped like this, I have no confidence the licence check, RUCSS/SaaS calls, and preload do not have equivalents.
7. **Post a public postmortem** covering why a blocking third-party call was ever allowed on the render path, and what now prevents a vendor endpoint from taking down customer sites again.

I have applied my own mitigation and my site is back to sub-second renders. I am posting this because most affected sites have no idea this is happening to them, and because the gap between "customer files a complete root-cause analysis" and "vendor ships a one-line fix" should not be measured in days.

**Comment by nightjoker7** [comment]
# Follow-up comment for wp-media/wp-rocket#8669

---

This is not a new failure mode for you. It is #6715, shipped again.

**#6715 — "Avoid requests to wp-rocket.me to negatively impact site performances during downtime"** (opened 2024-06-10, closed 2024-07-01, labelled `priority: 🔥critical`). Your own description of it:

> When WP Rocket (website) is down or slow, which can unfortunately happen, this creates slowness on every website checking for updates and license validity. The requests go in timeout (5 seconds).
>
> - Websites not in the cache are impacted
> - Admin is impacted
> - This potentially generates self-DDos
>
> **We've discussed this issue each time we had downtime without tackling it. It's time to tackle before another upcoming issue.**

That is a precise description of what is happening right now. You wrote it two years ago, marked it critical, and fixed it — for the licence check.

Then the analytics call shipped with the same blocking pattern and a **30-second** timeout. Six times the 5s you classified as critical. The mitigation you already built (timeout + circuit breaker) exists in this codebase. It was simply never generalised beyond the one endpoint that prompted it.

So the question is not "why did an endpoint go down." Endpoints go down. The question is why a known, documented, already-fixed failure mode was reintroduced on the render path with a worse timeout, and why the recurrence is being triaged more slowly than the original.

**#6715 was labelled `priority: 🔥critical`. #8669 currently has no label, no milestone, and no maintainer comment**, 27 hours after opening — while your own team posts internal one.com support tickets into this thread and three independent parties (myself, @gsusI on Cloudways/Nginx with New Relic traces showing a 77-second Store API transaction, and the original reporter) confirm the same call chain across different stacks.

Store API means carts and checkouts. This is taking revenue off paying customers' sites right now.

Two asks, both small:

1. **Label this `priority: 🔥critical`** to match #6715. The severity assessment is the thing blocking everything else.
2. **Ship `'blocking' => false` as a standalone hotfix today.** Not bundled into a feature release. Then do the circuit breaker and move analytics off the render path properly.

And a third, which matters more than the patch: **notify licence holders.** Most affected sites have no idea. They are debugging their own themes, blaming their hosts, and losing conversions to a defect in a plugin they pay you for specifically to make their sites faster.

**Comment by remyperona** [comment]
This is handled in the new version (1.4.2) of our Mixpanel library

**Comment by adamcapriola** [comment]
I've had to hard-disable WP Rocket since this issue essentially killed my website all day today. Waiting for a fix before I reactivate…

**Comment by jekayode** [comment]
Related case: https://onecomhelp.zendesk.com/agent/tickets/18242151

**Comment by DFC005** [comment]
This also sounds exactly like the issue I have. Rolling back to 3.22.1 makes my site fast and responsive again for logged in users but the latest version of the plugin is just unusable.

**Comment by DahmaniAdame** [comment]
We are working on a hotfix.

**Comment by hellofromahmed** [comment]
Released the fix, thanks everyone.
